library/ssl_tls.c

```diff
@@ -4474,17 +4470,34 @@ int mbedtls_ssl_parse_certificate( mbedtls_ssl_context *ssl )
 #endif /* MBEDTLS_ECP_C */
 
         if( mbedtls_ssl_check_cert_usage( ssl->session_negotiate->peer_cert,
-                                  ciphersuite_info,
-                                  ! ssl->conf->endpoint,
+                                 ciphersuite_info,
+                                 ! ssl->conf->endpoint,
                                  &ssl->session_negotiate->verify_result ) != 0 )
         {
             MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad certificate (usage extensions)" ) );
             if( ret == 0 )
                 ret = MBEDTLS_ERR_SSL_BAD_HS_CERTIFICATE;
         }
 
-        if( authmode == MBEDTLS_SSL_VERIFY_OPTIONAL )
+        /* mbedtls_x509_crt_verify_with_profile is supposed to report a
+         * verification failure through MBEDTLS_ERR_X509_CERT_VERIFY_FAILED,
+         * with details encoded in the verification flags. All other kinds
+         * of error codes, including those from the user provided f_vrfy
+         * functions, are treated as fatal and lead to a failure of
+         * ssl_parse_certificate even if verification was optional. */
+        if( authmode == MBEDTLS_SSL_VERIFY_OPTIONAL &&
+            ( ret == MBEDTLS_ERR_X509_CERT_VERIFY_FAILED ||
+              ret == MBEDTLS_ERR_SSL_BAD_HS_CERTIFICATE ) )
+        {
             ret = 0;
+        }
+
+        if( ca_chain == NULL && authmode == MBEDTLS_SSL_VERIFY_REQUIRED )
```

Comment:
Regardless of `f_vrfy` being set or not, the code will try to build up a trusted chain from the peer's cert to one of the trusted certs and e.g. check their validity and signatures. Moreover it will call `f_vrfy` for each of these certificates in case `f_vrfy` is present, the purpose of which is to do additional checks or to even clear flags the users doesn't care about. If `ca_chain` is `NULL`, at least the flag `X509_BADCERT_NOT_TRUSTED` by our code, but you're right in that in principle, the user might even remove this flag. So I agree that removing the `ca_chain` check should be ok, and even in line with `f_vrfy` being able to guard the verification - on the other hand, I'm a reluctant to deviate too much from the previous behaviour.